packages/headless/src/controllers/recent-queries-list/headless-recent-queries-list.ts

```diff
@@ -0,0 +1,148 @@
+import {updateQuery} from '../../features/query/query-actions';
+import {
+  logClearRecentQueries,
+  logRecentQueryClick,
+} from '../../features/recent-queries/recent-queries-analytics-actions';
+
+export interface RecentQueriesListProps {
+  /**
+   * The initial state that should be applied to the `RecentQueriesList` controller.
+   */
+  initialState?: RecentQueriesListInitialState;
+  /**
+   * The configuration options that should be applied to the `RecentQueriesList` controller.
+   */
+  options?: RecentQueriesListOptions;
+}
+
+export interface RecentQueriesListInitialState {
+  /**
+   * The list of recent queries
+   * @defaultValue `[]`
+   */
+  queries: string[];
+}
+
+const initialStateSchema = new Schema<RecentQueriesListInitialState>({
+  queries: new ArrayValue({required: false}),
+});
+
+export interface RecentQueriesListOptions {
+  /**
+   * The maximum number of queries to retain in the list.
+   * @defaultValue `10`
+   */
+  maxQueries: number;
+}
+
+const optionsSchema = new Schema<RecentQueriesListOptions>({
+  maxQueries: new NumberValue({required: false}),
+});
+
+/**
+ * The `RecentQueriesList` controller is in charge of managing the user's recent queries.
+ */
+export interface RecentQueriesList extends Controller {
+  /**
+   * The state of the RecentQueriesList controller.
+   * */
+  state: RecentQueriesState;
+  /**
+   * Clear the recent queries list.
+   */
+  clear(): void;
+  /**
+   * Execute the given recent query.
+   * @param query - The recent query to execute.
+   */
+  executeRecentQuery(query: string): void;
+}
+
+export function validateRecentQueriesProps(
+  engine: SearchEngine,
+  props?: RecentQueriesListProps
+) {
+  validateOptions(
+    engine,
+    optionsSchema,
+    props?.options,
+    'buildRecentQueriesList'
+  );
+  validateInitialState(
+    engine,
+    initialStateSchema,
+    props?.initialState,
+    'buildRecentQueriesList'
+  );
+}
+
+/**
+ * Creates a `RecentQueriesList` controller instance.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configuration `RecentQueriesList` properties.
+ * @returns A `RecentQueriesList` controller instance.
+ * */
+export function buildRecentQueriesList(
+  engine: SearchEngine,
+  props?: RecentQueriesListProps
+): RecentQueriesList {
+  if (!loadRecentQueriesListReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const controller = buildController(engine);
+  const {dispatch} = engine;
+  const getState = () => engine.state;
+
+  validateRecentQueriesProps(engine, props);
+  dispatch(
+    registerRecentQueries({
+      queries: props?.initialState?.queries ?? [],
+      maxQueries: props?.options?.maxQueries ?? 10,
+    })
+  );
+
+  return {
+    ...controller,
+
+    get state() {
+      const state = getState();
+
+      return state.recentQueries;
+    },
+
+    clear() {
+      dispatch(logClearRecentQueries());
+      dispatch(clearRecentQueries());
+    },
+
+    executeRecentQuery(value: string) {
```

Comment:
Should this really be `any value` versus `a value that must be contained in the array of recent queries` ?

Because then, it's not really `executeRecentQuery`, it is more `executeAnyQuery` ?

Reply:
fair enough!

Reply:
We could use an index

Reply:
An index would work indeed, just need to add some validation that it's not out of bound/that is'a really a number (using bueno, again).

Those validation are very important for developers using JavaScript instead of TypeScript, where they do not have access to the TS compiler telling them they are passing in the right type.

So they get a nice "clean" error message versus a cryptic stack trace.

Reply:
Understood. Thanks for including the rationale as well 👍